docs: Automatic Code Documentation

victorgarcia98 opened the pull request:
### Description

This PR makes use of the Sphinx extension `autosummary` to create code documentation straight out of the docstrings!

### Look & Feel
![Screenshot 2023-05-24 at 17-04-05 flexmeasures — FlexMeasures 0 14 dev16 documentation](https://github.com/FlexMeasures/flexmeasures/assets/84775131/7bb221f6-e351-48ad-9bbe-f01fd191e9fe)



### Testing Steps

```
python3 -m venv venv
source venv/bin/activate
make install
make update-docs
```

### Potential Improvements
- Deactivate automatic generation on every push as it's very time consuming. By toggling off `autosummary_generate`, for example via a env variable, we can switch off the docs generation. 

- Edit templates.

Closes #52.

nhoening reviewed (commented):
Cool stuff - much better than before, obviously!

First notes:
- This makes the building docs take longer, let's pay attention to that. Is there an easy way to skip this auto-documentation, with a flag or so?
- [possible follow-up issue] We could now fix some modules which have no docstring line in their `__init__.py`. Only utils has one. Same one level deeper.
- What is the difference between [this](flexmeasures/ui/static/documentation/html/_autosummary/flexmeasures.utils.config_utils.html#flexmeasures.utils.config_utils.read_config) and [this](flexmeasures/ui/static/documentation/html/_autosummary/flexmeasures.utils.config_utils.read_config.html), information-wise?

victorgarcia98 commented:
> This makes the building docs take longer, let's pay attention to that. Is there an easy way to skip this auto-documentation, with a flag or so?

I've added the flag `gen_code_docs` which controls if we want to generate the docs or not.

> What is the difference between [this](https://github.com/FlexMeasures/flexmeasures/pull/flexmeasures/ui/static/documentation/html/_autosummary/flexmeasures.utils.config_utils.html#flexmeasures.utils.config_utils.read_config) and [this](https://github.com/FlexMeasures/flexmeasures/pull/flexmeasures/ui/static/documentation/html/_autosummary/flexmeasures.utils.config_utils.read_config.html), information-wise?

You're right! I've simplified the docs to show the description of the objects (classes, functions, attributes and exceptions) in the same page, without any list. 

Other than that, we need to decided on the following topics:
- Format of the documentation for a (sub)module (order of elements, lists, ...).
- Modules to avoid, for example, the CLI it's pretty much covered in [this](https://flexmeasures.readthedocs.io/en/latest/cli/commands.html) page and autodoc fails to capture *Click commands*.
- To have or not have a main page for the code documentation.

Among them, I think some of these could be included in this PR and others might be included later.

nhoening commented:
Good improvements!

I think it feels weird right now that only "flexmeasures" is listed at first. Maybe it should get its own page, that would be nicer, and maybe it's the quickest way.

I would include the short docstrings for the main modules in this PR. Should not be more than ten minutes of work. I'll gladly do it :)

victorgarcia98 commented:
Thanks!

> I think it feels weird right now that only "flexmeasures" is listed at first. Maybe it should get its own page, that would be nicer, and maybe it's the quickest way.

For this, we can list the main modules when calling autocode.

> I would include the short docstrings for the main modules in this PR. Should not be more than ten minutes of work. I'll gladly do it :)

Okay, thanks. Feel free to do it in this branch or create a new PR to this branch, instead :smile:

nhoening reviewed (changes requested):
I feel the autosummary step is happening regardless of my usage of `gen_code_docs`.

At least, I have a length step of "reading sources..."

I tried this, but it didn't seem to work:

```
if autosummary_generate:
    extensions.add("sphinx.ext.autosummary")
    extensions.add("sphinx.ext.autodoc.typehints")
```
 
Maybe the `.. autosummary::` directive will make it happen regardless?

Also, look at [this view](flexmeasures/ui/static/documentation/html/_autosummary/flexmeasures.api.dev.sensors.AssetAPI.html#flexmeasures.api.dev.sensors.AssetAPI) which should only list the `AssetAPI.get` function (endpoint), but seems to list functions of `FlaskView`.

victorgarcia98 commented:
Thanks for trying it out :smile: 

> I feel the autosummary step is happening regardless of my usage of gen_code_docs.
> At least, I have a length step of "reading sources..."

The `gen_code_docs` actually disables the generation of the file structure, all the RST files containing the documentation for the modules (autosummary). This flag doesn't deactivate the generation of the HTML from the code (i.e autodoc).

I ran some tests to check the time of the building process in my local setup:

1) Benchmark (commenting the autosummary code in `index.rst` and `_autosummary` folder deleted):

real	0m12.756s
user	0m14.160s
sys	0m5.360s

real	0m12.878s
user	0m14.142s
sys	0m5.422s

2) Full code docs generation: `make update-docs gen_code_docs=True`

real	0m45.264s
user	0m46.346s
sys	0m5.641s

real	0m45.483s
user	0m46.203s
sys	0m5.513s

3) Disable autosummary  `make update-docs gen_code_docs=False`

real	0m45.039s
user	0m46.026s
sys	0m5.650s

real	0m46.185s
user	0m47.098s
sys	0m5.693s


Digging up a little bit, I found that the problem is that autodoc doesn't provide the intermediate RST files, but only the ouputs such as HTML or Latex. For that reason, 159 files need to be "compiled" internally to RST and then to the output format, which slows down the process.

Apparently, this is an open [Issue](https://github.com/sphinx-doc/sphinx/issues/2953) in the autodoc repo, but it looks It hasn't received much attention :cry:. Also, there is people showing interest in this in a [SO post](https://stackoverflow.com/questions/10377576/emit-restructuredtext-from-sphinx-autodoc), but the workarounds feel very hacky to me atm. 

For this reason, I would generate the documentation only when doing changes to main (merge or push) and I don't think it will be a problem in terms of build times given that it runs in 46s in my setup (please, find ReadTheDocs build limits [here](https://docs.readthedocs.io/en/stable/builds.html#build-resources))

> Also, look at [this view](https://github.com/FlexMeasures/flexmeasures/pull/flexmeasures/ui/static/documentation/html/_autosummary/flexmeasures.api.dev.sensors.AssetAPI.html#flexmeasures.api.dev.sensors.AssetAPI) which should only list the AssetAPI.get function (endpoint), but seems to list functions of FlaskView.

With the new change, I can't see this page being generated.  

In that page, I see the following:

![image](https://github.com/FlexMeasures/flexmeasures/assets/84775131/49b146f3-abe8-4d9f-b8d6-d9e49859bde6)

nhoening commented:
> With the new change, I can't see this page being generated.

I'll have to try again. Which new change do you mean?

Happy to see that on RTD we won't have a problem.

victorgarcia98 commented on documentation/conf.py:
```diff
@@ -52,11 +54,20 @@
 # Add any paths that contain templates here, relative to this directory.
 templates_path = ["_templates"]
 
+# generate autosummary even if no references
```
Not sure what this means, It's a left-over comment from where I copied this from :S

victorgarcia98 commented:
Good idea!

Running `make update-docs` or `make update-docs-pdf` set `gen_code_docs` to False, by default.

Otherwise, If they ENV variable `GEN_CODE_DOCS` is not found, the default would be `True`.

nhoening commented:
I think the Makefile should delete the documentation/_autosummary folder, as its existence leads to sphinx working on its content, no matter what `gen_code_docs` says.

victorgarcia98 commented:
I've added two commits to:
* delete the autosummary folder if `gen_code_docs = False`
* list the main FM modules under *CODE DOCUMENTATION*

nhoening reviewed (changes requested):
Hm, [something is off at ReadTheDocs now](https://readthedocs.org/projects/flexmeasures/builds/20949149/), we need to add some settings so the app creation runs through:

```
Running Sphinx v6.1.3
making output directory... done
[FLEXMEASURES][2023-06-07 21:39:52,388] INFO: Loading config from /home/docs/checkouts/readthedocs.org/user_builds/flexmeasures/envs/latest/var/flexmeasures-instance/flexmeasures.cfg ...
[FLEXMEASURES][2023-06-07 21:39:52,389] WARNING: File /home/docs/checkouts/readthedocs.org/user_builds/flexmeasures/envs/latest/var/flexmeasures-instance/flexmeasures.cfg could not be found! (work dir is /home/docs/checkouts/readthedocs.org/user_builds/flexmeasures/checkouts/latest/documentation)
Missing the required configuration settings: SQLALCHEMY_DATABASE_URI
You can provide these settings ― as environment variables or in your config file (e.g. /home/docs/.flexmeasures.cfg or /home/docs/checkouts/readthedocs.org/user_builds/flexmeasures/envs/latest/var/flexmeasures-instance/flexmeasures.cfg).
```

Or we just need to set `app.env` correctly, so [this line](https://github.com/FlexMeasures/flexmeasures/blob/main/flexmeasures/utils/config_utils.py#L97) will work...

nhoening commented:
It did? In only see failed attempts here: https://readthedocs.org/projects/flexmeasures/builds/

victorgarcia98 commented:
Oh, I meant locally (with the secret key set)

nhoening commented:
I don't recommend setting those in the Makefile, as ReadTheDocs is not using that :) Also then these stick in the devs environment and have to be unset.
See how I am adding them in .readthedocs.yaml. However, I also failed with SECRET_KEY. Will go to bed now...

victorgarcia98 commented:
I've tested just to set the variables `SQLALCHEMY_DB_URI` and `SECRET_KEY` in `documentation/conf.py` and the build [succeded](https://readthedocs.org/projects/flexmeasures/builds/20952438/). You can see the results [here](https://flexmeasures.readthedocs.io/en/latest/).

Please, let me know if there's any other request from your end. Otherwise, we can merge :smile:

nhoening commented:
As I mentioned earlier, setting these in conf.py might influence the developer environment.
Aren't you overwriting your FM configuration with dummy values, making it impossible to test real functionality?

nhoening commented on flexmeasures/utils/config_utils.py:
```diff
@@ -81,7 +81,7 @@ def read_config(app: Flask, custom_path_to_config: str | None):
     path_to_config_instance = os.path.join(app.instance_path, "flexmeasures.cfg")
 
-    # Custom config: not when testing (that should run completely on defaults)
+    # Custom config: not when testing (that should run completely on defaults) and documentation
```
Either remove documentation from the comment as well, or re-add it as a test condition. I have a feeling it might be needed.

victorgarcia98 commented:
> Thanks, it works indeed!
> 
> The docs building (HTML & PDF) both went from ~40 seconds to ~140 seconds in RTD. Seems we can afford it. The overall build time has other overhead, as well.

Great!! Merging into main :fireworks:

Flix6x commented:
Congrats!

We did forget the changelog entry that this deserves. Let's just leave the "Still Needs Changelog Entry" label so we catch it during our release process.

nhoening commented:
I added the entry, but the internet in the train did not work at that moment, apparently. I'll add it to main now.
